File: backend/src/application/guides/guides.service.ts
import { Injectable, NotFoundException } from '@nestjs/common'
import { PrismaService } from '../../infra/prisma/prisma.service'
import type { CategoryId, GuideDto, PlaceCardDto } from '../../contracts/dto/place.dto'
import { PLACE_INCLUDE, toCard, type PlaceWithRelations } from '../places/place.mapper'

const FREE_GUIDE_ID = 'free-to-play'

/* Mirrors buildGuides() in the mock: one guide per category with >=2 spots,
   plus a "Free to play" cross-cut prepended. Derived from the place set. */
@Injectable()
export class GuidesService {
  constructor(private readonly prisma: PrismaService) {}

  private async loadPlaces(): Promise<PlaceWithRelations[]> {
    return this.prisma.place.findMany({ include: PLACE_INCLUDE, orderBy: { sort: 'asc' } })
  }

  private catId(p: PlaceWithRelations): CategoryId {
    const link = p.categories.find((pc) => pc.primary) ?? p.categories[0]
    return link.category.id as CategoryId
  }

  private async build(): Promise<{ guides: GuideDto[]; places: PlaceWithRelations[] }> {
    const [cats, places] = await Promise.all([
      this.prisma.category.findMany({ orderBy: { sort: 'asc' } }),
      this.loadPlaces(),
    ])

    const byCat: GuideDto[] = cats
      .map((c) => {
        const slugs = places.filter((p) => this.catId(p) === c.id).map((p) => p.slug)
        return {
          id: `cat-${c.id}`,
          title: c.label,
          subtitle: `Every ${c.short.toLowerCase()} spot in the field guide`,
          coverCategory: c.id as CategoryId,
          count: slugs.length,
          spotSlugs: slugs,
        }
      })
      .filter((g) => g.count >= 2)

    const freeSlugs = places.filter((p) => p.isFree).map((p) => p.slug)
    const free: GuideDto = {
      id: FREE_GUIDE_ID,
      title: 'Free to play',
      subtitle: 'No booking, no fee — just show up',
      coverCategory: 'scenic',
      count: freeSlugs.length,
      spotSlugs: freeSlugs,
    }

    return { guides: [free, ...byCat], places }
  }

  async list(): Promise<GuideDto[]> {
    return (await this.build()).guides
  }

  async getById(id: string): Promise<{ guide: GuideDto; spots: PlaceCardDto[] }> {
    const { guides, places } = await this.build()
    const guide = guides.find((g) => g.id === id)
    if (!guide) throw new NotFoundException(`guide not found: ${id}`)
    const bySlug = new Map(places.map((p) => [p.slug, p]))
    const spots = guide.spotSlugs
      .map((s) => bySlug.get(s))
      .filter((p): p is PlaceWithRelations => !!p)
      .map(toCard)
    return { guide, spots }
  }
}

File: api/src/main/java/ee/gemspot/api/service/GuidesService.java
package ee.gemspot.api.service;

import ee.gemspot.api.domain.Category;
import ee.gemspot.api.domain.Place;
import ee.gemspot.api.dto.GuideDetailDto;
import ee.gemspot.api.dto.GuideDto;
import ee.gemspot.api.dto.PlaceCardDto;
import ee.gemspot.api.mapper.PlaceMapper;
import ee.gemspot.api.repository.CategoryRepository;
import ee.gemspot.api.repository.PlaceRepository;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.server.ResponseStatusException;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Mirrors buildGuides() in the mock: one guide per category with >=2 spots,
 * plus a "Free to play" cross-cut prepended. Derived from the place set.
 */
@Service
@Transactional(readOnly = true)
public class GuidesService {

    private static final String FREE_GUIDE_ID = "free-to-play";

    private final PlaceRepository placeRepository;
    private final CategoryRepository categoryRepository;
    private final PlaceMapper mapper;

    public GuidesService(PlaceRepository placeRepository,
                         CategoryRepository categoryRepository,
                         PlaceMapper mapper) {
        this.placeRepository = placeRepository;
        this.categoryRepository = categoryRepository;
        this.mapper = mapper;
    }

    private String catId(Place p) {
        return mapper.primaryCategory(p).getId();
    }

    /** Build result mirroring Nest's { guides, places } so getById does not re-query. */
    private record Built(List<GuideDto> guides, List<Place> places) {}

    private Built build() {
        List<Category> cats = categoryRepository.findAllByOrderBySortAsc();
        List<Place> places = placeRepository.findAllByOrderBySortAsc();

        List<GuideDto> byCat = new ArrayList<>();
        for (Category c : cats) {
            List<String> slugs = places.stream()
                    .filter(p -> catId(p).equals(c.getId()))
                    .map(Place::getSlug)
                    .toList();
            GuideDto guide = new GuideDto(
                    "cat-" + c.getId(),
                    c.getLabel(),
                    "Every " + c.getShortLabel().toLowerCase() + " spot in the field guide",
                    c.getId(),
                    null,
                    slugs.size(),
                    slugs
            );
            if (guide.count() >= 2) byCat.add(guide);
        }

        List<String> freeSlugs = places.stream()
                .filter(Place::isFree)
                .map(Place::getSlug)
                .toList();
        GuideDto free = new GuideDto(
                FREE_GUIDE_ID,
                "Free to play",
                "No booking, no fee — just show up",
                "scenic",
                null,
                freeSlugs.size(),
                freeSlugs
        );

        List<GuideDto> guides = new ArrayList<>();
        guides.add(free);
        guides.addAll(byCat);
        return new Built(guides, places);
    }

    public List<GuideDto> list() {
        return build().guides();
    }

    public GuideDetailDto getById(String id) {
        Built built = build();
        GuideDto guide = built.guides().stream()
                .filter(g -> g.id().equals(id))
                .findFirst()
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "guide not found: " + id));

        Map<String, Place> bySlug = new LinkedHashMap<>();
        for (Place p : built.places()) bySlug.put(p.getSlug(), p);

        List<PlaceCardDto> spots = guide.spotSlugs().stream()
                .map(bySlug::get)
                .filter(p -> p != null)
                .map(mapper::toCard)
                .toList();

        return new GuideDetailDto(guide, spots);
    }
}
